src/canister/user_index/src/api/canister_management/recycle_canisters.rs

```diff
@@ -0,0 +1,118 @@
+            canister_data
+                .wasms
+                .get(&WasmType::IndividualUserWasm)
+                .unwrap()
+                .clone()
+        });
+
+        let futures = canister_ids.iter().map(|canister_id| {
+            recycle_canister(*canister_id, individual_user_template_canister_wasm.clone())
+        });
+
+        let stream = futures::stream::iter(futures).boxed().buffer_unordered(20);
+
+        let results = stream
+            .collect::<Vec<Result<Option<Principal>, (Principal, String)>>>()
+            .await;
+
+        // update recycle_status
+
+        let success_canisters = results
+            .iter()
+            .filter(|r| r.is_ok() && r.as_ref().unwrap().is_some())
+            .map(|r| r.as_ref().unwrap().unwrap())
+            .collect::<Vec<Principal>>();
+
+        let num_success = success_canisters.len();
+
+        CANISTER_DATA.with_borrow_mut(|canister_data| {
+            canister_data
+                .available_canisters
+                .extend(success_canisters.clone());
+
+            // remove the canisters that are recycled from user_principal_id_to_canister_id_map and unique_user_name_to_user_principal_id_map
+            // canister_id is the value in the map
+            for canister_id in success_canisters.clone() {
+                canister_data
+                    .user_principal_id_to_canister_id_map
+                    .retain(|_, v| *v != canister_id);
+                canister_data
+                    .unique_user_name_to_user_principal_id_map
+                    .retain(|_, v| *v != canister_id);
+            }
+        });
+
+        let failed_list = results
+            .iter()
+            .filter(|r| r.is_err())
+            .map(|r| r.as_ref().unwrap_err().clone())
+            .collect::<Vec<(Principal, String)>>();
+
+        let end = get_current_system_time();
+
+        CANISTER_DATA.with_borrow_mut(|canister_data| {
+            canister_data.recycle_status = RecycleStatus {
+                success_canisters: success_canisters.iter().map(|c| c.to_string()).collect(),
+                last_recycled_at: Some(end),
+                last_recycled_duration: Some(end.duration_since(start).unwrap().as_secs()),
+                num_last_recycled_canisters: num_success as u64,
+                failed_recycling: failed_list,
+            };
+        });
+    });
+}
+
+pub async fn recycle_canister(
+    canister_id: Principal,
+    individual_user_template_canister_wasm: CanisterWasm,
+) -> Result<Option<Principal>, (Principal, String)> {
+    // reinstall wasm
+    match reinstall_canister_wasm(
+        canister_id,
+        None,
+        individual_user_template_canister_wasm.version.clone(),
+        individual_user_template_canister_wasm.wasm_blob.clone(),
+    )
+    .await
+    {
+        Ok(_) => Ok(Some(canister_id)),
+        Err(e) => Err((canister_id, e)),
+    }
+}
```

Comment:
This is same as `reset_user_individual_canisters`. See if we can refactor some code and use common functions.